openid-4-verifiable-presentations-1_0.md

```diff
@@ -549,11 +555,276 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+no specific constraints are placed on the metadata or validity of the requested
+Credential.
+
+`expect_claims`:
+: OPTIONAL. A non-empty array of objects as defined in (#claims_query) that specifies
+claims in the requested Credentials.
+
+`valid_claim_sets`:
+: OPTIONAL. A non-empty array containing arrays of identifiers for elements in
+`expect_claims`. The identifier MAY be postfixed by `?`, indicating that
+delivery of the the respective claim is OPTIONAL.  
+
+## Claims Query {#claims_query}
+
+Each entry in `expect_claims` MUST be an object with the following properties:
+
+`id`:
+: REQUIRED if `valid_claim_sets` is present; OPTIONAL otherwise. A string
+identifying the particular claim. The value MUST be a non-empty string
+consisting of alphanumeric, underscore (`_`) or hyphen (`-`) characters.
+Within the particular `expected_claims` array, the same `id` MUST NOT
+be present more than once.
+
+`path`:
+: REQUIRED if the Credential Format uses a JSON-based claims structure; MUST NOT
+be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
+within the Verifiable Credential, as defined in (#claims_path_pointer).
+
+`namespace`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the namespace of the claim
+within the Verifiable Credential, e.g., `org.iso.18013.5.1`.
+
+`claim_name`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the name of the claim within the provided namespace
+in the Verifiable Credential, e.g., `first_name`.
+
+`value`:
+: OPTIONAL. A string, integer or boolean value that specifies the expected value of the claim. If the
+`value` property is present, the Wallet MUST return the claim only if the type and value
+of the claim match the type and value specified in the query.
+
+`values`:
+: OPTIONAL. An array of strings, integers or boolean values that specifies the expected values of the claim.
+If the `values` property is present, the Wallet MUST return the claim only if the
+type and value of the claim both match for at least one of the elements in the array.
+
+### Selecting Claims and Credentials {#vp_query_lang_processing_rules}
+
+The same basic logic applies for selecting claims and for selecting credentials, as detailed in the following.
+
+#### Selecting Claims
+
+The following rules apply for selecting claims via `expect_claims` and `valid_claim_sets`:
+
+- If `expect_claims` is not provided, the Verifier requests all claims existing
+  in the Credential.
+- If `expect_claims` is provided, but `valid_claim_sets` is not provided,
+  the Verifier requests all claims listed in `expect_claims`.
+- Otherwise, the Verifier requests one combination of the claims listed in
```

Comment:
Is there a way to indicate preference for a certain claim? (typically for privacy reason, e.g. give me age_21 OR date_of_birth, preferring age_21)

Reply:
Not yet - it could be implicit by the order in `valid_claim_sets`

Reply:
+1 to this as a feature and using simple array ordering to express preference would appear to be pragmatic.

Reply:
Either we use the existing order in `valid_claim_sets` or we could change this structure...

```json
  "valid_credential_sets": [
    ["pid", "nice_to_have?"],
    ["other_pid", "nice_to_have?"],
    ["pid_reduced_cred_1", "pid_reduced_cred_2", "nice_to_have?"]
  ]
```

to use objects in the array:

```json
  "valid_credential_sets": [
    {"priority": 3, "set": ["pid", "nice_to_have?"]},
    {"priority": 1, "set": ["other_pid", "nice_to_have?"]},
    {"priority": 1, "set": ["pid_reduced_cred_1", "pid_reduced_cred_2", "nice_to_have?"]}
  ]
```

Reply:
I think using order in the array is sufficient. a better example is probably

"valid_credential_sets": [
    ["pid_reduced_cred_1?", "pid_reduced_cred_2?",]
  ]

where give me pid_reduced_cred_1 or pid_reduced_cred_2 but 1 is preferred

Reply:
Is it then the intent to have order of preference for both claims as well as credentials?

The assumption is that when there are choices (give me this claim or that claim, give me this credential set or that credential set), there may be some privacy or usability benefits to making a particular choice, but both of them will satisfy the request of the reader and therefore give you the service, satisfy the use case etc.
This is why we say it's a preference, not a requirement. 

For claims, having an order of preference makes a lot of sense (hence the comment) since its boundary is within the credential. Typically the order of preference will/should be used for privacy reasons (e.g. age_over_18 vs date_of_birth, city_name vs full_address). There should be little to no cases where not meeting the order of preference from the reader comes at a cost (typically usability) to the holder.

But for credentials, there are a lot more considerations as to which of the options to satisfy. Some examples:
- Privacy (credential x is more minimized than credential y)
- Wallets preference by the user (credential x requires wallet x, but credential y can be used with wallet y)
- User authentication differences (credential x may be more cumbersome than credential y)
- UX (how do you indicate in a meaningful way that an RP prefers credential x, but credential y will still be acceptable)
- Presentation time (credential x may reside remote and therefore be slower than credential y)

Given that it ultimately shouldn't matter to the RP which credential is returned. I think it's safer to not have an order of preference for credentials for now, as different parts of the ecosystem (wallets, credential selector, RPs) can interpret the meaning of 'preference' differently which can lead to confusion.
We can then revisit this when we have more experience with implementations and real-life scenarios as to how to best approach this.

Reply:
this got addressed by changing credentials to array of objects from array of arrays.